src/planner/select_plan.go

```diff
@@ -186,20 +189,14 @@ func (p *SelectPlan) Build() error {
 		}
 	}
 
+	expr, _ := t.tableExpr.Expr.(sqlparser.TableName)
 	// Rewritten the query.
 	for _, segment := range segments {
-		tn := sqlparser.TableName{
-			Name:      sqlparser.NewTableIdent(segment.Table),
-			Qualifier: sqlparser.NewTableIdent(shardDatabase),
-		}
-		as := fmt.Sprintf(" as %s", aliasTable)
+		// Rewrite the shard table's name
+		expr.Name = sqlparser.NewTableIdent(segment.Table)
+		t.tableExpr.Expr = expr
 		buf := sqlparser.NewTrackedBuffer(nil)
-		buf.Myprintf("select %v%s%v from %v%s%v%v%v%v%v",
-			node.Comments, node.Hints, node.SelectExprs,
-			tn, as,
-			node.Where,
-			node.GroupBy, node.Having, node.OrderBy,
-			node.Limit)
+		node.Format(buf)
```

Comment:
After the last Iterate, the expr.Name of the ast will be changed to the segment.Table of the last, 
the ast should be const?

Reply:
I can add the code to recovery the expr.Name.
 `expr.Name = sqlparser.NewTableIdent(t.tableName)`
 ` t.tableExpr.Expr = expr`
But I think that is unnecessary. The ast `node` is a pointer, used to build plan tree. After the distributed querys build, the ast is useless.  The ast is also changed when creating sub-plans such as aggregate, orderby, limit .